File: packages/protocol/src/policy.ts
import type { FeedPost, FeedPostKind, PendingAction, ReviewState, TaskCommandState } from "./index.js";

// Shared feed / outbox / reconnect policy. These rules used to be handwritten
// in apps/web (TypeScript) and apps/ios (Swift) and started to drift; every
// client must now implement this module exactly. packages/protocol/test-vectors
// holds the versioned JSON cases that both the vitest and XCTest suites assert.

// --- Feed ---

export const FEED_MERGE_WINDOW_MS = 6 * 60 * 60 * 1_000;

export const POST_VALUE: Record<FeedPostKind, number> = {
  failure: 1,
  result: 2,
  decision: 3,
  attention: 3,
  progress: 4,
};

// Only progress/decision posts merge; progress/decision/attention posts sink
// once a later outcome covers them.
const ROUTINE_POST_KINDS: readonly FeedPostKind[] = ["progress", "decision"];
const COVERABLE_POST_KINDS: readonly FeedPostKind[] = ["progress", "decision", "attention"];

const COVERED_PRIORITY_PENALTY = 6;
const READ_PRIORITY_PENALTY = 10;

// Merges routine posts newest-first: a progress/decision post folds into the
// newest post with the same `${sessionId ?? taskId}:${kind}` key when it is at
// most FEED_MERGE_WINDOW_MS older, appending deduplicated highlights capped at
// two. The comparison is always against the newest post of the key, so merged
// chains do not extend the window.
export function mergeRoutinePosts(posts: FeedPost[]): FeedPost[] {
  const merged: FeedPost[] = [];
  const latestByKey = new Map<string, number>();
  for (const post of [...posts].sort((left, right) => right.createdAt.localeCompare(left.createdAt))) {
    if (!ROUTINE_POST_KINDS.includes(post.kind)) {
      merged.push(post);
      continue;
    }
    const key = `${post.sessionId ?? post.taskId}:${post.kind}`;
    const existingIndex = latestByKey.get(key);
    const existing = existingIndex === undefined ? undefined : merged[existingIndex];
    const withinWindow = existing && new Date(existing.createdAt).getTime() - new Date(post.createdAt).getTime() <= FEED_MERGE_WINDOW_MS;
    if (!existing || !withinWindow) {
      latestByKey.set(key, merged.length);
      merged.push(post);
      continue;
    }
    merged[existingIndex!] = { ...existing, highlights: [...existing.highlights, ...post.highlights].filter((value, index, all) => all.indexOf(value) === index).slice(0, 2) };
  }
  return merged;
}

export function postNeedsAction(input: {
  actionRequired: boolean;
  hasLinkedPendingAction: boolean;
  directReplyIsCurrent: boolean;
  reviewState: ReviewState | null;
}): boolean {
  return input.reviewState === "unreviewed"
    || (input.actionRequired && (input.hasLinkedPendingAction || input.directReplyIsCurrent));
}

// A coverable post is covered when the task's newest result/failure outcome is
// strictly newer than the post.
export function isPostCovered(input: {
  kind: FeedPostKind;
  createdAt: string;
  latestOutcomeCreatedAt: string | null;
}): boolean {
  return COVERABLE_POST_KINDS.includes(input.kind)
    && (input.latestOutcomeCreatedAt ?? "") > input.createdAt;
}

export function postPriority(input: {
  kind: FeedPostKind;
  needsAction: boolean;
  covered: boolean;
  unread: boolean;
}): number {
  if (input.needsAction) return 0;
  return POST_VALUE[input.kind]
    + (input.covered ? COVERED_PRIORITY_PENALTY : 0)
    + (input.unread ? 0 : READ_PRIORITY_PENALTY);
}

export interface FeedOrderable {
  priority: number;
  createdAt: string;
}

// Ascending priority, then newest first. Feed building sorts with a stable
// sort, so items equal on both keys keep their input order.
export function compareFeedItems(left: FeedOrderable, right: FeedOrderable): number {
  return left.priority - right.priority || right.createdAt.localeCompare(left.createdAt);
}

// --- Command outbox semantic keys ---
//
// Dedupe keys for durable outbox commands, as the union of the web
// (commandOutbox.ts) and iOS (AppModel.swift) rules:
// - every string field is trimmed before joining (iOS trimmed all fields, web
//   only free text; ids are machine-generated, so trimming changes nothing)
// - action.decide sorts its input record by key (code-unit order, not locale)
//   so JSON field order never changes the key
// - feed.seen / task.timeline.seen keys come from the iOS rule set, the
//   action.decide sorted-input rule from web
// - unknown command types fall back to a recursively key-sorted JSON encoding
//   of the command fields (type excluded, it is already the key prefix),
//   replacing web's insertion-order JSON.stringify and iOS's random UUID so
//   the fallback is deterministic too

export type SemanticCommandFields = { type: string } & Record<string, unknown>;

export function semanticCommandKey(command: SemanticCommandFields): string {
  switch (command.type) {
    case "task.create":
      return `${command.type}:${textField(command, "provider")}:${textField(command, "workspaceId")}:${textField(command, "text")}`;
    case "task.follow_up":
    case "session.message":
      return `${command.type}:${textField(command, "sessionId")}:${textField(command, "text")}`;
    case "task.command.retry":
      return `${command.type}:${textField(command, "commandId")}`;
    case "task.command.cancel":
      return `${command.type}:${textField(command, "commandId") || textField(command, "idempotencyKey")}`;
    case "action.decide":
      return `${command.type}:${textField(command, "actionId")}:${textField(command, "decisionId")}:${textField(command, "confirmationPhrase")}:${sortedInput(command.input)}`;
    case "feed.dismiss":
    case "feed.dismiss.set":
      return `${command.type}:${textField(command, "itemId")}`;
    case "feed.seen":
      return `${command.type}:${textField(command, "postId")}`;
    case "task.timeline.seen":
      return `${command.type}:${textField(command, "sessionId")}:${textField(command, "itemId")}`;
    case "agent.profile.update":
    case "trust.policy.update":
      return `${command.type}:${textField(command, "projectId")}`;
    case "user.profile.update":
    case "notification.settings.update":
    case "notification.device.register":
    case "notification.device.unregister":
      return command.type;
    case "review.respond":
      return `${command.type}:${textField(command, "reviewId")}:${textField(command, "decision")}:${textField(command, "note")}`;
    default: {
      const { type: _, ...fields } = command;
      return `${command.type}:${stableStringify(fields)}`;
    }
  }
}

function textField(command: Record<string, unknown>, key: string): string {
  const value = command[key];
  return typeof value === "string" ? value.trim() : "";
}

function sortedInput(input: unknown): string {
  if (!input || typeof input !== "object" || Array.isArray(input)) return "[]";
  const entries = Object.entries(input as Record<string, unknown>)
    .filter((entry): entry is [string, string] => typeof entry[1] === "string")
    .sort(([left], [right]) => compareCodeUnits(left, right));
  return JSON.stringify(entries);
}

function stableStringify(value: unknown): string {
  if (Array.isArray(value)) return `[${value.map((item) => stableStringify(item)).join(",")}]`;
  if (value && typeof value === "object") {
    const entries = Object.entries(value as Record<string, unknown>)
      .filter(([, entryValue]) => entryValue !== undefined)
      .sort(([left], [right]) => compareCodeUnits(left, right));
    return `{${entries.map(([key, entryValue]) => `${JSON.stringify(key)}:${stableStringify(entryValue)}`).join(",")}}`;
  }
  return JSON.stringify(value) ?? "null";
}

function compareCodeUnits(left: string, right: string): number {
  return left < right ? -1 : left > right ? 1 : 0;
}

// --- Command cancellation ---

// Only queued commands may be canceled; dispatching/running already reached the
// agent and the terminal states are settled.
export const CANCELABLE_COMMAND_STATES: readonly TaskCommandState[] = ["queued"];

export function isCommandCancelable(state: TaskCommandState): boolean {
  return CANCELABLE_COMMAND_STATES.includes(state);
}

// --- Reconnect backoff ---

export const BACKOFF_DELAYS_MS: readonly number[] = [1_000, 2_000, 4_000, 8_000, 16_000, 30_000];

export const BACKOFF_JITTER_RATIO = 0.2;

// attempt 0 is the first retry; attempts past the schedule clamp to the last
// delay. A symmetric ±20% jitter (injectable random for tests) keeps
// reconnecting clients from synchronizing.
export function backoffDelayMs(attempt: number, random: () => number = Math.random): number {
  const index = Math.min(Math.max(Math.trunc(attempt), 0), BACKOFF_DELAYS_MS.length - 1);
  const base = BACKOFF_DELAYS_MS[index]!;
  return Math.round(base * (1 + (random() * 2 - 1) * BACKOFF_JITTER_RATIO));
}

// --- Quick approval ---

// A pending action qualifies for one-tap approval from a notification or feed
// card only when it is an approval offering an explicit low-risk allow-once
// decision and a deny decision, neither requiring a confirmation phrase.
export function isQuickApprovable(action: Pick<PendingAction, "kind" | "availableDecisions">): boolean {
  if (action.kind !== "approval") return false;
  const allowOnce = action.availableDecisions.find((decision) => decision.scope === "once");
  const deny = action.availableDecisions.find((decision) => decision.scope === "deny");
  return Boolean(
    allowOnce && allowOnce.risk === "low" && !allowOnce.confirmationPhrase
    && deny && !deny.confirmationPhrase,
  );
}

File: runtime/crates/zimlo-protocol/src/policy.rs
use std::{cmp::Ordering, collections::HashMap};

use chrono::DateTime;
use serde::{Deserialize, Serialize};
use serde_json::{Map, Value};

pub const FEED_MERGE_WINDOW_MS: i64 = 6 * 60 * 60 * 1_000;
pub const COVERED_PRIORITY_PENALTY: i32 = 6;
pub const READ_PRIORITY_PENALTY: i32 = 10;
pub const BACKOFF_DELAYS_MS: [u64; 6] = [1_000, 2_000, 4_000, 8_000, 16_000, 30_000];
pub const BACKOFF_JITTER_RATIO: f64 = 0.2;
pub const CANCELABLE_COMMAND_STATES: [&str; 1] = ["queued"];

#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "lowercase")]
pub enum FeedPostKind {
    Failure,
    Result,
    Decision,
    Attention,
    Progress,
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct FeedPostSummary {
    pub id: String,
    pub kind: FeedPostKind,
    pub task_id: String,
    pub session_id: Option<String>,
    pub created_at: String,
    pub highlights: Vec<String>,
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct FeedOrderable {
    pub id: String,
    pub priority: i32,
    pub created_at: String,
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct DecisionSummary {
    pub scope: String,
    pub risk: String,
    #[serde(default)]
    pub confirmation_phrase: Option<String>,
}

pub fn merge_routine_posts(posts: &[FeedPostSummary]) -> Vec<FeedPostSummary> {
    let mut sorted = posts.to_vec();
    sorted.sort_by(|left, right| right.created_at.cmp(&left.created_at));

    let mut merged: Vec<FeedPostSummary> = Vec::new();
    let mut latest_by_key: HashMap<String, usize> = HashMap::new();
    for post in sorted {
        if !matches!(post.kind, FeedPostKind::Progress | FeedPostKind::Decision) {
            merged.push(post);
            continue;
        }

        let owner = post.session_id.as_deref().unwrap_or(&post.task_id);
        let key = format!("{owner}:{:?}", post.kind);
        let Some(existing_index) = latest_by_key.get(&key).copied() else {
            latest_by_key.insert(key, merged.len());
            merged.push(post);
            continue;
        };

        let within_window =
            milliseconds_between(&merged[existing_index].created_at, &post.created_at)
                .is_some_and(|difference| difference <= FEED_MERGE_WINDOW_MS);
        if !within_window {
            latest_by_key.insert(key, merged.len());
            merged.push(post);
            continue;
        }

        for highlight in post.highlights {
            if merged[existing_index].highlights.len() >= 2 {
                break;
            }
            if !merged[existing_index].highlights.contains(&highlight) {
                merged[existing_index].highlights.push(highlight);
            }
        }
    }
    merged
}

pub fn is_post_covered(
    kind: FeedPostKind,
    created_at: &str,
    latest_outcome_created_at: Option<&str>,
) -> bool {
    matches!(
        kind,
        FeedPostKind::Progress | FeedPostKind::Decision | FeedPostKind::Attention
    ) && latest_outcome_created_at.is_some_and(|outcome| outcome > created_at)
}

pub fn post_priority(kind: FeedPostKind, needs_action: bool, covered: bool, unread: bool) -> i32 {
    if needs_action {
        return 0;
    }
    post_value(kind)
        + if covered { COVERED_PRIORITY_PENALTY } else { 0 }
        + if unread { 0 } else { READ_PRIORITY_PENALTY }
}

pub fn compare_feed_items(left: &FeedOrderable, right: &FeedOrderable) -> Ordering {
    left.priority
        .cmp(&right.priority)
        .then_with(|| right.created_at.cmp(&left.created_at))
}

pub fn semantic_command_key(command: &Value) -> String {
    let Some(command) = command.as_object() else {
        return String::new();
    };
    let command_type = text_field(command, "type");
    let host = text_field(command, "hostId");
    let local_key = match command_type.as_str() {
        "task.create" => with_material_ids(
            format!(
                "{}:{}:{}:{}",
                command_type,
                text_field(command, "provider"),
                text_field(command, "workspaceId"),
                text_field(command, "text")
            ),
            command.get("materialIds"),
        ),
        "task.follow_up" | "session.message" => with_material_ids(
            format!(
                "{}:{}:{}",
                command_type,
                text_field(command, "sessionId"),
                text_field(command, "text")
            ),
            command.get("materialIds"),
        ),
        "material.register" => format!(
            "{}:{}",
            command_type,
            command
                .get("material")
                .and_then(Value::as_object)
                .map_or_else(String::new, |material| text_field(material, "id"))
        ),
        "task.command.retry" => {
            format!("{}:{}", command_type, text_field(command, "commandId"))
        }
        "task.command.cancel" => {
            let command_id = text_field(command, "commandId");
            let identifier = if command_id.is_empty() {
                text_field(command, "idempotencyKey")
            } else {
                command_id
            };
            format!("{command_type}:{identifier}")
        }
        "action.decide" => format!(
            "{}:{}:{}:{}:{}",
            command_type,
            text_field(command, "actionId"),
            text_field(command, "decisionId"),
            text_field(command, "confirmationPhrase"),
            sorted_input(command.get("input"))
        ),
        "feed.dismiss" | "feed.dismiss.set" => {
            format!("{}:{}", command_type, text_field(command, "itemId"))
        }
        "feed.seen" => format!("{}:{}", command_type, text_field(command, "postId")),
        "task.timeline.seen" => format!(
            "{}:{}:{}",
            command_type,
            text_field(command, "sessionId"),
            text_field(command, "itemId")
        ),
        "agent.profile.update" | "trust.policy.update" => {
            format!("{}:{}", command_type, text_field(command, "projectId"))
        }
        "user.profile.update"
        | "notification.settings.update"
        | "notification.device.register"
        | "notification.device.unregister" => command_type.clone(),
        _ => {
            let fields = command
                .iter()
                .filter(|(key, _)| key.as_str() != "type" && key.as_str() != "hostId")
                .map(|(key, value)| (key.clone(), value.clone()))
                .collect::<Map<String, Value>>();
            format!(
                "{}:{}",
                command_type,
                stable_stringify(&Value::Object(fields))
            )
        }
    };
    if host.is_empty() {
        local_key
    } else {
        format!("{host}:{local_key}")
    }
}

pub fn is_command_cancelable(state: &str) -> bool {
    CANCELABLE_COMMAND_STATES.contains(&state)
}

pub fn backoff_delay_ms(attempt: f64, random_value: f64) -> u64 {
    let index = (attempt.trunc() as isize).clamp(0, BACKOFF_DELAYS_MS.len() as isize - 1) as usize;
    let base = BACKOFF_DELAYS_MS[index] as f64;
    (base * (1.0 + (random_value * 2.0 - 1.0) * BACKOFF_JITTER_RATIO)).round() as u64
}

pub fn is_quick_approvable(kind: &str, decisions: &[DecisionSummary]) -> bool {
    if kind != "approval" {
        return false;
    }
    let allow_once = decisions.iter().find(|decision| decision.scope == "once");
    let deny = decisions.iter().find(|decision| decision.scope == "deny");
    allow_once
        .is_some_and(|decision| decision.risk == "low" && decision.confirmation_phrase.is_none())
        && deny.is_some_and(|decision| decision.confirmation_phrase.is_none())
}

fn post_value(kind: FeedPostKind) -> i32 {
    match kind {
        FeedPostKind::Failure => 1,
        FeedPostKind::Result => 2,
        FeedPostKind::Decision | FeedPostKind::Attention => 3,
        FeedPostKind::Progress => 4,
    }
}

fn milliseconds_between(newer: &str, older: &str) -> Option<i64> {
    let newer = DateTime::parse_from_rfc3339(newer).ok()?;
    let older = DateTime::parse_from_rfc3339(older).ok()?;
    Some((newer - older).num_milliseconds())
}

fn text_field(command: &Map<String, Value>, key: &str) -> String {
    command
        .get(key)
        .and_then(Value::as_str)
        .map(str::trim)
        .unwrap_or_default()
        .to_owned()
}

fn with_material_ids(base: String, value: Option<&Value>) -> String {
    match value.and_then(Value::as_array) {
        Some(materials) if !materials.is_empty() => {
            format!(
                "{base}:{}",
                stable_stringify(&Value::Array(materials.clone()))
            )
        }
        _ => base,
    }
}

fn sorted_input(value: Option<&Value>) -> String {
    let Some(input) = value.and_then(Value::as_object) else {
        return "[]".to_owned();
    };
    let mut entries = input
        .iter()
        .filter_map(|(key, value)| value.as_str().map(|value| (key, value)))
        .collect::<Vec<_>>();
    entries.sort_by(|(left, _), (right, _)| compare_code_units(left, right));
    serde_json::to_string(&entries).expect("string pairs always serialize")
}

fn stable_stringify(value: &Value) -> String {
    match value {
        Value::Array(values) => format!(
            "[{}]",
            values
                .iter()
                .map(stable_stringify)
                .collect::<Vec<_>>()
                .join(",")
        ),
        Value::Object(object) => {
            let mut entries = object.iter().collect::<Vec<_>>();
            entries.sort_by(|(left, _), (right, _)| compare_code_units(left, right));
            let fields = entries
                .into_iter()
                .map(|(key, value)| {
                    format!(
                        "{}:{}",
                        serde_json::to_string(key).expect("JSON object keys always serialize"),
                        stable_stringify(value)
                    )
                })
                .collect::<Vec<_>>()
                .join(",");
            format!("{{{fields}}}")
        }
        _ => serde_json::to_string(value).expect("JSON values always serialize"),
    }
}

fn compare_code_units(left: &str, right: &str) -> Ordering {
    left.encode_utf16().cmp(right.encode_utf16())
}
